internal/monitors/device-updater/device_updater.go

```diff
@@ -77,8 +83,76 @@ func (d *DeviceUpdater) updateDeviceSpecAccordingToFleetTemplate(log logrus.Fiel
 	device.Generation = &targetGeneration
 	device.Spec = model.MakeJSONField(fleet.Spec.Data.Template.Spec)
 	log.Infof("Updating device %s to generation %d", device.Name, *device.Generation)
-	err := d.devStore.UpdateDeviceInternal(device)
-	if err != nil {
-		log.Errorf("failed to update target generation for device %s (fleet %s): %w", device.Name, fleet.Name, err)
+}
+
+func (d *DeviceUpdater) updateDeviceHeartbeatConditions(log logrus.FieldLogger, device *model.Device, fleet *model.Fleet) {
+	// If the device has never updated its status, there's nothing to do
+	if device.Status.Data.UpdatedAt == nil {
+		return
+	}
+
+	warningTime := fleet.Spec.Data.DeviceConditions.HeartbeatElapsedTimeWarning
+	errorTime := fleet.Spec.Data.DeviceConditions.HeartbeatElapsedTimeError
+	now := util.TimeStampStringPtr()
+
+	if device.Status.Data.Conditions == nil {
+		device.Status.Data.Conditions = &[]api.DeviceCondition{
```

Comment:
just so I understand in terms of ownership how are we handling who can write to status. Like this appears an external monitor is updating the status of a device vs the device itself? Or is the device making a call to update status and this is the result.

Reply:
The device calls the updatestatus API, but some conditions have to be updated by the service logic